src/session.coffee

```diff
@@ -3,77 +3,131 @@ define [
   'promise'
   'runnable'
   'biskoto'
-  'session_engines/get_param_engine'
   'session_engines/xdomain_engine'
-], (Settings, Promise, Runnable, Biskoto, GetParamEngine, XDomainEngine) ->
+  'analytics_url'
+  'domain_extractor'
+], (Settings, Promise, Runnable, Biskoto, XDomainEngine, AnalyticsUrl, DomainExtractor) ->
   class Session
     Session::[key] = method for key, method of Runnable
 
     constructor: (@plugins_manager) ->
       @promise = new Promise()
       @_cleanUpCookies()
-      @analytics_session = @_getCookieAnalyticsSession()
+      @analytics_session = @_getSessionFromCookie()
 
     then: (success, fail) -> @promise.then(success, fail)
 
     _commands:
       session:
         create: (shop_code, flavor, metadata) ->
           @shop_code = shop_code
+          @flavor = flavor
           @metadata = metadata
 
-          @_extractAnalyticsSession('create', shop_code, flavor, encodeURIComponent(JSON.stringify(metadata)))
+          @_extractAnalyticsSessionOnCreate()
 
         connect: (shop_code)->
           # connect should be called only once
           return console?.warn?('Connect called multiple times') if @shop_code
 
           @shop_code = shop_code
-          @plugins_manager.session = @
+          @metadata = @_getMetadataFromCookie()
+          @domain = new DomainExtractor(Settings.url.hostname).get()
 
+          @plugins_manager.session = @
           @plugins_manager.notify('connect', {})
 
-          return @promise.resolve(@) if @analytics_session
+          skr_prm = new AnalyticsUrl(Settings.url.current).read_params()
+          @_setAnalyticsSession(skr_prm)
+          @_setAnalyticsMetadata(skr_prm)
+
+          return @promise.resolve(@) if @analytics_session || @analytics_session = @_getSessionFromCacheCookie()
```

Comment:
I think it would be more readable:

```coffeescript
@analytics_session ||= @_getSessionFromCacheCookie()
return @promise.resolve(@) if @analytics_session
```